cpp/ql/src/experimental/Likely Bugs/OverrunWriteProductFlow.ql

```diff
@@ -78,6 +78,141 @@ predicate isSinkPairImpl(
+      )
+    }
+
+    predicate includeHiddenNodes() { any() }
+  }
+
+  private import DataFlow::GlobalWithState<ValidStateConfig>
+
+  private predicate inLoop(PathNode n) { n.getASuccessor+() = n }
+
+  /**
+   * Holds if `value` is a possible offset for `n`.
+   *
+   * To ensure termination, we limit `value` to be in the
+   * range `[-2, 2]` if the node is part of a loop. Without
+   * this restriction we wouldn't terminate on an example like:
+   * ```cpp
+   * while(unknown()) { size++; }
+   * ```
+   */
+  private predicate validStateImpl(PathNode n, int value) {
+    // If the dataflow node depends recursively on itself we restrict the range.
+    (inLoop(n) implies value = [-2 .. 2]) and
+    (
+      // For the dataflow source we have an allocation such as `malloc(size + k)`,
+      // and the value of the flow-state is then `k`.
+      hasSize(_, n.getNode(), value)
+      or
+      // For a dataflow sink any `value` that is strictly smaller than the delta
+      // needs to be a valid flow-state. That is, for a snippet like:
+      // ```
+      // p = new char[size];
+      // memset(p, 0, size + 2);
+      // ```
+      // the valid flow-states at the `memset` must include `0` since the flow-state
+      // at the source is `0`. Similarly, for an example such as:
+      // ```
+      // p = new char[size + 1];
+      // memset(p, 0, size + 2);
+      // ```
+      // the flow-state at the `memset` must include `1` since `1` is the flow-state
+      // after the source.
+      //
+      // So we find a valid flow-state at the sink's predecessor, and use the definition
+      // of our sink predicate to compute the valid flow-states at the sink.
+      exists(int delta, PathNode n0 |
+        n0.getASuccessor() = n and
+        validStateImpl(n0, value) and
+        isSinkPairImpl(_, _, n.getNode(), delta, _) and
+        delta > value
+      )
+      or
+      // For a non-source and non-sink node there is two cases to consider.
+      // 1. A node where we have to update the flow-state, or
+      // 2. A node that doesn't update the flow-state.
+      //
+      // For case 1, we compute the new flow-state by adding the constant operand of the
+      // `AddInstruction` to the flow-state of any predecessor node.
+      // For case 2 we simply propagate the valid flow-states from the predecessor node to
+      // the next one.
+      exists(PathNode n0, DataFlow::Node node, int value0 |
+        n0.getASuccessor() = n and
+        validStateImpl(n0, value0) and
+        node = n.getNode()
+      |
+        exists(AddInstruction add, Operand op1, Operand op2, int delta |
+          add = node.asInstruction() and
+          add.hasOperands(op1, op2) and
+          value0 = value + delta and
+          semBounded(getSemanticExpr([op1, op2].getDef()), any(SemZeroBound zero), delta, true, _)
+        )
+        or
+        not node.asInstruction() instanceof AddInstruction and
+        value = value0
+      )
```

Comment:
To me this feels like special-casing something the range analysis library should already been doing. Why Isn't is sufficient to use the `hasSize` predicate?

Reply:
This is traversing the entire interprocedural path. I don't think this is something we can ever hope to do with the range analysis library.

Reply:
f1c124a3da3ffd8e1d91cae95fd6d779b7481c9d hopefully makes this predicate (and its relationship with the product-flow configuration) slightly more clear.

Reply:
This has been resolved. This code should match that in the dataflow configuration, which it does now.